[torch.futures] Add note about error handling for non-chained futures.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53212 [torch.futures] Add note about error handling for non-chained futures.**

Ran into a strange issue with error handling in future callbacks, more
details in https://github.com/pytorch/pytorch/issues/52166, but essentially,
after a callback throws all additional processing stops, and other futures can
never be completed, resulting in a hang. Add a note to warn about this.

In particular, the following can result in a hang:

```
    ret_fut = torch.futures.Future9)
    def test_rpc_err_in_cb(self):
        if self.rank != 0:
            return
        # this will cause the RPC to be marked with an error due to timeout. 
        fut = rpc.rpc_async(worker_name(self.rank + 1), slow_add, args=(torch.tensor(1), torch.tensor(1)), timeout=0.01)
        def cb(fut):
            print(f"Fut result is {fut.wait()}") # throws
            ret_fut.set_result(fut.wait())

        fut.add_done_callback(cb)
        ret_fut.wait()
```

Differential Revision: [D26793310](https://our.internmc.facebook.com/intern/diff/D26793310/)

rohan-varma committed:
[torch.futures] Add note about error handling for non-chained futures.

Ran into a strange issue with error handling in future callbacks, more
details in https://github.com/pytorch/pytorch/issues/52132, but essentially,
after a callback throws all additional processing stops, and other futures can
never be completed, resulting in a hang. Add a note to warn about this.

Differential Revision: [D26793310](https://our.internmc.facebook.com/intern/diff/D26793310/)

[ghstack-poisoned]

SciPioneer commented:
I have only used `then()`. Don't quite understand why when we should prefer `add_done_callback` to `then`.

rohan-varma commented:
> I have only used `then()`. Don't quite understand when we should prefer `add_done_callback` to `then`.

One use case I had is found in https://github.com/pytorch/pytorch/pull/52166. In that PR i used `add_done_callback` to mark another future as completed as I wasn't able to use `then()` due to the API surface.

SciPioneer commented on torch/futures/__init__.py:
```diff
@@ -61,6 +61,15 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
+            through the original future being completed with an exception and
+            calling ``fut.wait()``, or through other code in the callback, the
+            future returned by ``then`` will be marked appropriately with the
+            encountered error. However, if this callback later completes
```
Probably better to mention this extra error handling is currently only needed by `add_done_callback()`, not `then()`?

rohan-varma replied:
This can also happen with `then`, i.e. if users use `then` to schedule arbitrary callbacks. This is possible since `_add_done_callback` is currently a private api and not exposed. i.e., in the example in the PR, the same issue is possible with `then()`.

rohan-varma committed:
Update on "[torch.futures] Add note about error handling for non-chained futures."


Ran into a strange issue with error handling in future callbacks, more
details in https://github.com/pytorch/pytorch/issues/52166, but essentially,
after a callback throws all additional processing stops, and other futures can
never be completed, resulting in a hang. Add a note to warn about this.

In particular, the following can result in a hang:

```
    ret_fut = torch.futures.Future9)
    def test_rpc_err_in_cb(self):
        if self.rank != 0:
            return
        # this will cause the RPC to be marked with an error due to timeout. 
        fut = rpc.rpc_async(worker_name(self.rank + 1), slow_add, args=(torch.tensor(1), torch.tensor(1)), timeout=0.01)
        def cb(fut):
            print(f"Fut result is {fut.wait()}") # throws
            ret_fut.set_result(fut.wait())

        fut.add_done_callback(cb)
        ret_fut.wait()
```

Differential Revision: [D26793310](https://our.internmc.facebook.com/intern/diff/D26793310/)

[ghstack-poisoned]
